### ISSUE TITLE
Pass complete request to catchers, for easier request replay

### ISSUE THREAD

**Issue by c-eliasson** [elbywan/wretch#15]
Just started using wretch and I really like the API. Currently building a large app, with a micro service architecture, and really appreciate how you can reuse a wretch object for each micro service component, with very small variations for each endpoint.

One thing that I'm missing, that would be convenient, is to have the complete wretch object for a given request, passed as an argument to the callback of any catcher.

The scenario I'm trying to address is primarily HTTP 401 responses, where my access token have expired. In that case, I would like to have a callback that can refresh the token, and then take the failed request and replay it (but with a new access token).

**Pseudo-code example:**

```
wretch("...")
  .get()
  .unauthorized((err, req) => {
    const token = getNewToken();
    // Replay the failed request, with the new token
    const response = req.auth(`Bearer ${token}`).json()
  })
```

_Not the most useful example, but I believe it shows what I'm after._

What do you think about that?

**Comment by elbywan** [comment]
@c-eliasson

> Just started using wretch and I really like the API.

Thanks a bunch 👍 !

> What do you think about that?

Actually I have the exact same use case in the project I'm currently working on !

Just made a [commit](https://github.com/elbywan/wretch/commit/c10eb422dd67eccdb8d8ac88b40fa12b5c474b94) on the dev branch which should alleviate the issue by passing the original request along for catcher/resolver callbacks.

Here's an example for early catchers :

```js
const reAuthOn401 = wretch()
  .catcher(401, (error, originalRequest) => {
    // Renew credentials
    const token = getToken()
    // Replay the original request with new credentials
    return originalRequest
      .auth(`Bearer ${ token }`)
      .get()
      // Redefine authorized to prevent infinite loops in case or multiple 401 errors
      .unauthorized(err => { throw err })
      .json()
  })

reAuthOn401.url("/resource")
  .get()
  .json() // <- Will only be called for the original promise
  .then(callback) // <- Will be called for the original OR the replayed promise result
```

And for regular error catchers :

```js
wretch("/resource")
  .get()
  .unauthorized((error, req) => {
    const token = getToken()
    return originalRequest
      .auth(`Bearer ${ token }`)
      .get()
      .unauthorized(err => { throw err })
      .json()
  })
  .json()
  .then(callback)
```

Could you check this out and tell me what you think about the syntax and if it solves your issue ? 

Thanks !

**Comment by c-eliasson** [comment]
Wow, quick response indeed! This looks exactly like what I'm after, very nice! :+1:

**Comment by elbywan** [comment]
@c-eliasson Released with version 1.1.1 📦
